File: frontend/apps/ios/Pantopus/Features/DiscoverHub/DiscoverHubSampleData.swift
//
//  DiscoverHubSampleData.swift
//  Pantopus
//
//  A11.3 Discover magazine sample data. Backend is intentionally not used
//  for this surface; the view-model seeds deterministic content so previews,
//  app renders, and snapshot tests stay stable.
//

import Foundation
import SwiftUI

public enum DiscoverHubMagazineScenario: Sendable, Hashable {
    case loading
    case empty
    case populated
    case error
}

public enum DiscoverHubMagazineState: Sendable, Hashable {
    case loading
    case empty
    case populated(DiscoverHubMagazineContent)
    case error(message: String)
}

public struct DiscoverHubMagazineContent: Sendable, Hashable {
    public let pins: [DiscoverHubMapPin]
    public let cluster: DiscoverHubMapCluster
    public let tasks: [DiscoverHubTaskCard]
    public let marketplace: [DiscoverHubMarketplaceCard]
    public let posts: [DiscoverHubPostCard]
}

public enum DiscoverHubMapKind: String, CaseIterable, Sendable, Hashable {
    case task
    case item
    case post
    case spot
    case event

    public var label: String {
        switch self {
        case .task: "Task"
        case .item: "Item"
        case .post: "Post"
        case .spot: "Spot"
        case .event: "Event"
        }
    }

    public var pluralLabel: String {
        switch self {
        case .task: "Tasks"
        case .item: "Items"
        case .post: "Posts"
        case .spot: "Spots"
        case .event: "Events"
        }
    }

    public var color: Color {
        switch self {
        case .task: Theme.Color.handyman
        case .item: Theme.Color.goods
        case .post: Theme.Color.primary500
        case .spot: Theme.Color.home
        case .event: Theme.Color.business
        }
    }

    public var softColor: Color {
        switch self {
        case .task: Theme.Color.warningBg
        case .item: Theme.Color.businessBg
        case .post: Theme.Color.primary50
        case .spot: Theme.Color.homeBg
        case .event: Theme.Color.businessBg
        }
    }

    public var icon: PantopusIcon {
        switch self {
        case .task: .hammer
        case .item: .tag
        case .post: .messageCircle
        case .spot: .shoppingBag
        case .event: .calendar
        }
    }
}

public struct DiscoverHubMapPin: Identifiable, Sendable, Hashable {
    public let id: String
    public let kind: DiscoverHubMapKind
    public let x: CGFloat
    public let y: CGFloat
    public let pulses: Bool

    public init(
        id: String,
        kind: DiscoverHubMapKind,
        x: CGFloat,
        y: CGFloat,
        pulses: Bool = false
    ) {
        self.id = id
        self.kind = kind
        self.x = x
        self.y = y
        self.pulses = pulses
    }
}

public struct DiscoverHubMapCluster: Sendable, Hashable {
    public let count: Int
    public let x: CGFloat
    public let y: CGFloat
}

public struct DiscoverHubTaskCard: Identifiable, Sendable, Hashable {
    public let id: String
    public let title: String
    public let price: String
    public let distance: String
    public let bids: String
}

public struct DiscoverHubMarketplaceCard: Identifiable, Sendable, Hashable {
    public let id: String
    public let title: String
    public let price: String
    public let distance: String
    public let icon: PantopusIcon
}

public struct DiscoverHubPostCard: Identifiable, Sendable, Hashable {
    public let id: String
    public let intent: String
    public let title: String
    public let body: String
    public let author: String
    public let replies: Int
}

public enum DiscoverHubSampleData {
    public static let populated = DiscoverHubMagazineContent(
        pins: [
            DiscoverHubMapPin(id: "pin-task-1", kind: .task, x: 0.19, y: 0.53, pulses: true),
            DiscoverHubMapPin(id: "pin-item-1", kind: .item, x: 0.42, y: 0.26),
            DiscoverHubMapPin(id: "pin-spot-1", kind: .spot, x: 0.63, y: 0.63),
            DiscoverHubMapPin(id: "pin-post-1", kind: .post, x: 0.82, y: 0.39),
            DiscoverHubMapPin(id: "pin-event-1", kind: .event, x: 0.14, y: 0.84),
            DiscoverHubMapPin(id: "pin-task-2", kind: .task, x: 0.86, y: 0.84)
        ],
        cluster: DiscoverHubMapCluster(count: 9, x: 0.56, y: 0.82),
        tasks: [
            DiscoverHubTaskCard(
                id: "gig-shelves",
                title: "Hang 3 floating shelves",
                price: "$60",
                distance: "0.2 mi",
                bids: "4 bids"
            ),
            DiscoverHubTaskCard(
                id: "gig-clean",
                title: "Deep clean 2BR before move-out",
                price: "$180",
                distance: "0.5 mi",
                bids: "7 bids"
            ),
            DiscoverHubTaskCard(
                id: "gig-dog-walks",
                title: "Midday dog walks Tue/Thu",
                price: "$22/walk",
                distance: "0.3 mi",
                bids: "2 bids"
            )
        ],
        marketplace: [
            DiscoverHubMarketplaceCard(
                id: "item-sideboard",
                title: "Mid-century walnut sideboard",
                price: "$420",
                distance: "0.4 mi",
                icon: .home
            ),
            DiscoverHubMarketplaceCard(
                id: "item-bike",
                title: "Vintage Trek road bike, 56cm",
                price: "$240",
                distance: "0.7 mi",
                icon: .package
            ),
            DiscoverHubMarketplaceCard(
                id: "item-skillet",
                title: "Cast-iron Lodge skillet set",
                price: "$45",
                distance: "0.2 mi",
                icon: .utensils
            ),
            DiscoverHubMarketplaceCard(
                id: "item-kallax",
                title: "Ikea Kallax 4x4, white",
                price: "Free",
                distance: "0.6 mi",
                icon: .archive
            )
        ],
        posts: [
            DiscoverHubPostCard(
                id: "post-cardiologist",
                intent: "Ask",
                title: "Anyone know a good cardiologist nearby?",
                body: "Specifically near Hayes Valley - recently moved and looking for someone who takes Blue Shield.",
                author: "Maya - 0.1 mi",
                replies: 8
            ),
            DiscoverHubPostCard(
                id: "post-ramen",
                intent: "Recommend",
                title: "The new ramen place on Divisadero is fantastic",
                body: "Went last night. Tonkotsu was excellent. They open at 5 and the line moves fast.",
                author: "Dre - 0.3 mi",
                replies: 14
            )
        ]
    )

    public static let emptySkeletonRailTitles = [
        "Tasks near you",
        "Marketplace picks"
    ]
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/discoverhub/DiscoverHubSampleData.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.ui.screens.discoverhub

import androidx.compose.runtime.Immutable
import androidx.compose.ui.graphics.Color
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon

/**
 * A11.3 Discover magazine sample data. Backend is intentionally not used
 * for this surface; the view-model seeds deterministic content so previews,
 * app renders, and snapshot tests stay stable.
 */
enum class DiscoverHubMagazineScenario {
    Loading,
    Empty,
    Populated,
    Error,
}

sealed interface DiscoverHubMagazineUiState {
    data object Loading : DiscoverHubMagazineUiState

    data object Empty : DiscoverHubMagazineUiState

    data class Populated(
        val content: DiscoverHubMagazineContent,
    ) : DiscoverHubMagazineUiState

    data class Error(
        val message: String,
    ) : DiscoverHubMagazineUiState
}

@Immutable
data class DiscoverHubMagazineContent(
    val pins: List<DiscoverHubMapPin>,
    val cluster: DiscoverHubMapCluster,
    val tasks: List<DiscoverHubTaskCard>,
    val marketplace: List<DiscoverHubMarketplaceCard>,
    val posts: List<DiscoverHubPostCard>,
)

enum class DiscoverHubMapKind(
    val key: String,
    val label: String,
    val pluralLabel: String,
    val color: Color,
    val softColor: Color,
    val icon: PantopusIcon,
) {
    Task(
        key = "task",
        label = "Task",
        pluralLabel = "Tasks",
        color = PantopusColors.handyman,
        softColor = PantopusColors.warningBg,
        icon = PantopusIcon.Hammer,
    ),
    Item(
        key = "item",
        label = "Item",
        pluralLabel = "Items",
        color = PantopusColors.goods,
        softColor = PantopusColors.businessBg,
        icon = PantopusIcon.Tag,
    ),
    Post(
        key = "post",
        label = "Post",
        pluralLabel = "Posts",
        color = PantopusColors.primary500,
        softColor = PantopusColors.primary50,
        icon = PantopusIcon.MessageCircle,
    ),
    Spot(
        key = "spot",
        label = "Spot",
        pluralLabel = "Spots",
        color = PantopusColors.home,
        softColor = PantopusColors.homeBg,
        icon = PantopusIcon.ShoppingBag,
    ),
    Event(
        key = "event",
        label = "Event",
        pluralLabel = "Events",
        color = PantopusColors.business,
        softColor = PantopusColors.businessBg,
        icon = PantopusIcon.Calendar,
    ),
}

@Immutable
data class DiscoverHubMapPin(
    val id: String,
    val kind: DiscoverHubMapKind,
    val x: Float,
    val y: Float,
    val pulses: Boolean = false,
)

@Immutable
data class DiscoverHubMapCluster(
    val count: Int,
    val x: Float,
    val y: Float,
)

@Immutable
data class DiscoverHubTaskCard(
    val id: String,
    val title: String,
    val price: String,
    val distance: String,
    val bids: String,
)

@Immutable
data class DiscoverHubMarketplaceCard(
    val id: String,
    val title: String,
    val price: String,
    val distance: String,
    val icon: PantopusIcon,
)

@Immutable
data class DiscoverHubPostCard(
    val id: String,
    val intent: String,
    val title: String,
    val body: String,
    val author: String,
    val replies: Int,
)

object DiscoverHubSampleData {
    val populated =
        DiscoverHubMagazineContent(
            pins =
                listOf(
                    DiscoverHubMapPin("pin-task-1", DiscoverHubMapKind.Task, x = 0.19f, y = 0.53f, pulses = true),
                    DiscoverHubMapPin("pin-item-1", DiscoverHubMapKind.Item, x = 0.42f, y = 0.26f),
                    DiscoverHubMapPin("pin-spot-1", DiscoverHubMapKind.Spot, x = 0.63f, y = 0.63f),
                    DiscoverHubMapPin("pin-post-1", DiscoverHubMapKind.Post, x = 0.82f, y = 0.39f),
                    DiscoverHubMapPin("pin-event-1", DiscoverHubMapKind.Event, x = 0.14f, y = 0.84f),
                    DiscoverHubMapPin("pin-task-2", DiscoverHubMapKind.Task, x = 0.86f, y = 0.84f),
                ),
            cluster = DiscoverHubMapCluster(count = 9, x = 0.56f, y = 0.82f),
            tasks =
                listOf(
                    DiscoverHubTaskCard("gig-shelves", "Hang 3 floating shelves", "$60", "0.2 mi", "4 bids"),
                    DiscoverHubTaskCard("gig-clean", "Deep clean 2BR before move-out", "$180", "0.5 mi", "7 bids"),
                    DiscoverHubTaskCard("gig-dog-walks", "Midday dog walks Tue/Thu", "$22/walk", "0.3 mi", "2 bids"),
                ),
            marketplace =
                listOf(
                    DiscoverHubMarketplaceCard("item-sideboard", "Mid-century walnut sideboard", "$420", "0.4 mi", PantopusIcon.Home),
                    DiscoverHubMarketplaceCard("item-bike", "Vintage Trek road bike, 56cm", "$240", "0.7 mi", PantopusIcon.Package),
                    DiscoverHubMarketplaceCard("item-skillet", "Cast-iron Lodge skillet set", "$45", "0.2 mi", PantopusIcon.Utensils),
                    DiscoverHubMarketplaceCard("item-kallax", "Ikea Kallax 4x4, white", "Free", "0.6 mi", PantopusIcon.Archive),
                ),
            posts =
                listOf(
                    DiscoverHubPostCard(
                        id = "post-cardiologist",
                        intent = "Ask",
                        title = "Anyone know a good cardiologist nearby?",
                        body = "Specifically near Hayes Valley - recently moved and looking for someone who takes Blue Shield.",
                        author = "Maya - 0.1 mi",
                        replies = 8,
                    ),
                    DiscoverHubPostCard(
                        id = "post-ramen",
                        intent = "Recommend",
                        title = "The new ramen place on Divisadero is fantastic",
                        body = "Went last night. Tonkotsu was excellent. They open at 5 and the line moves fast.",
                        author = "Dre - 0.3 mi",
                        replies = 14,
                    ),
                ),
        )

    val emptySkeletonRailTitles = listOf("Tasks near you", "Marketplace picks")
}
